### PR TITLE
Fix LeanFFI 

### PR DESCRIPTION
**Pull request by andrewmwells-amazon**
*Issue #, if available:*
https://github.com/cedar-policy/cedar-spec/issues/227

*Description of changes:*
~Fix LeanFFI to work with multiple threads.~ Also decrement reference count on responses to fix memory leak.

Threads issue is blocked on https://github.com/leanprover/lean4/pull/3632 being released. This PR now purely fixes the memory leak.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-drt/tests/integration_tests.rs
Comment (andrewmwells-amazon) [inline cedar-drt/tests/integration_tests.rs:104]: - [ ] Remove this before merge.

File: cedar-drt/tests/benchmark.rs
Comment (andrewmwells-amazon) [inline cedar-drt/tests/benchmark.rs:168]: - [ ] This is still broken.

**Comment by khieta** [comment]
Awesome progress! Thanks for looking into this 🙂

File: cedar-drt/tests/benchmark.rs
Comment (andrewmwells-amazon) [reply cedar-drt/tests/benchmark.rs]: Working (at least locally)

**Comment by andrewmwells-amazon** [comment]
The tests now run but Lean folks said this shouldn't work. Waiting for clarification on this thread: https://leanprover.zulipchat.com/#narrow/stream/424609-lean-at-aws/topic/Reverse.20FFI.20Thread.20Safety before we can merge.

**Comment by andrewmwells-amazon** [comment]
[Lean PR](https://github.com/leanprover/lean4/pull/3632). We need to wait until we upgrade to that Lean version before we merge this.

File: cedar-drt/src/lean_impl.rs
Comment (khieta) [inline cedar-drt/src/lean_impl.rs:214]: Is there a reason this code can't be in a `lean_obj_to_string` function like before? I feel like it'd be prettier to have it in one place (and will prevent me from forgetting `lean_dec` in the future 😉)

**Review by khieta — APPROVED** [review]

**Review by aaronjeline — APPROVED** [review]

**Pull request merged**
